### PR TITLE
Fix issue with recursive structures

### PR DESCRIPTION
**Pull request by nikolay-egorov**

### CODE REVIEW

File: jupyter-lib/api/src/main/kotlin/org/jetbrains/kotlinx/jupyter/api/VariableState.kt [hunk 1]
```diff
@@ -16,18 +18,81 @@ data class VariableStateImpl(
...
+        } else {
+            getRecursiveObjectCode()
+        }
+    }
+
+    private fun goThrough(value: Any?, seenNodes: MutableSet<Any>) {
+        if (value == null) return
+        val membersProperties = try {
+            value::class.declaredMemberProperties
+        } catch (e: Error) {
```
Comment (ileasile) [inline jupyter-lib/api/src/main/kotlin/org/jetbrains/kotlinx/jupyter/api/VariableState.kt:64]: Throwable?

File: jupyter-lib/api/src/main/kotlin/org/jetbrains/kotlinx/jupyter/api/VariableState.kt [hunk 2]
```diff
@@ -16,18 +18,81 @@ data class VariableStateImpl(
...
+
+        val receivedInstances: MutableList<Any?> = mutableListOf()
+        for (property in membersProperties) {
+            property as KProperty1<Any, *>
+            try {
+                val wasAccessible = property.isAccessible
+                property.isAccessible = true
+                val callInstance = property.get(value)
+                property.isAccessible = wasAccessible
+                val wasSeen = if (callInstance != null) !seenNodes.add(callInstance) else false
```
Comment (ileasile) [inline jupyter-lib/api/src/main/kotlin/org/jetbrains/kotlinx/jupyter/api/VariableState.kt:78]:
```
callInstance != null && !seenNodes.add(callInstance)
```

File: jupyter-lib/api/src/main/kotlin/org/jetbrains/kotlinx/jupyter/api/VariableState.kt [hunk 3]
```diff
@@ -16,18 +18,81 @@ data class VariableStateImpl(
...
     override val value: Any?
         get() = cachedValue
+
+    private fun handleRecursiveCase() {
+        if (cachedValue == null) return
+        val seenNodes = mutableSetOf<Any>()
+        goThrough(cachedValue, seenNodes)
+    }
+
+    private fun getProperString(): String? {
```
Comment (ileasile) [inline jupyter-lib/api/src/main/kotlin/org/jetbrains/kotlinx/jupyter/api/VariableState.kt:44]: It's not necessary to move this logic to a separate function, you can leave it in the getter

File: jupyter-lib/api/src/main/kotlin/org/jetbrains/kotlinx/jupyter/api/VariableState.kt [hunk 4]
```diff
@@ -16,18 +18,81 @@ data class VariableStateImpl(
...
+                property.isAccessible = wasAccessible
+                val wasSeen = if (callInstance != null) !seenNodes.add(callInstance) else false
+
+                if (wasSeen) {
+                    isRecursive = true
+                    return
+                }
+                receivedInstances.add(callInstance)
+            } catch (ex: Error) {
+                if (ex is StackOverflowError) {
```
Comment (ileasile) [inline jupyter-lib/api/src/main/kotlin/org/jetbrains/kotlinx/jupyter/api/VariableState.kt:86]: Either test or comment is needed

File: jupyter-lib/api/src/main/kotlin/org/jetbrains/kotlinx/jupyter/api/VariableState.kt [hunk 5]
```diff
@@ -16,18 +18,81 @@ data class VariableStateImpl(
...
     private var cachedValue: Any? = null
+    private var isRecursive: Boolean = false
 
     fun update() {
         val wasAccessible = property.isAccessible
         property.isAccessible = true
         val fieldValue = property.get(scriptInstance)
         property.isAccessible = wasAccessible
         cachedValue = fieldValue
+        handleRecursiveCase()
```
Comment (ileasile) [inline jupyter-lib/api/src/main/kotlin/org/jetbrains/kotlinx/jupyter/api/VariableState.kt:29]: checkIfRecursiveStructure?

File: jupyter-lib/api/src/main/kotlin/org/jetbrains/kotlinx/jupyter/api/VariableState.kt [hunk 6]
```diff
@@ -16,18 +18,81 @@ data class VariableStateImpl(
...
     override val stringValue: String?
-        get() = cachedValue?.toString()
+        get() = getProperString()
 
     override val value: Any?
         get() = cachedValue
+
+    private fun handleRecursiveCase() {
+        if (cachedValue == null) return
+        val seenNodes = mutableSetOf<Any>()
```
Comment (ileasile) [inline jupyter-lib/api/src/main/kotlin/org/jetbrains/kotlinx/jupyter/api/VariableState.kt:40]: Inline this local variable

File: jupyter-lib/api/src/main/kotlin/org/jetbrains/kotlinx/jupyter/api/VariableState.kt [hunk 7]
```diff
@@ -16,18 +18,81 @@ data class VariableStateImpl(
...
+        goThrough(cachedValue, seenNodes)
+    }
+
+    private fun getProperString(): String? {
+        fun getRecursiveObjectCode(): String {
+            val kClassName = cachedValue!!::class.simpleName
+            return buildString {
+                append("$kClassName: ")
+                append("recursive structure")
+            }
```
Comment (ileasile) [inline jupyter-lib/api/src/main/kotlin/org/jetbrains/kotlinx/jupyter/api/VariableState.kt:50]:
```suggestion
            return "$kClassName: recursive structure"
```

File: jupyter-lib/api/src/main/kotlin/org/jetbrains/kotlinx/jupyter/api/VariableState.kt [hunk 8]
```diff
@@ -16,18 +18,81 @@ data class VariableStateImpl(
...
         property.isAccessible = true
         val fieldValue = property.get(scriptInstance)
         property.isAccessible = wasAccessible
         cachedValue = fieldValue
+        handleRecursiveCase()
     }
 
     override val stringValue: String?
-        get() = cachedValue?.toString()
+        get() = getProperString()
```
Comment (ileasile) [inline jupyter-lib/api/src/main/kotlin/org/jetbrains/kotlinx/jupyter/api/VariableState.kt:33]: Maybe `by lazy` will be good here?

**Review by ileasile — CHANGES_REQUESTED** [review]
Let's also make the serialization switched off by default. Let's do it only if it was requested by execute request (can we make an additional optional field in it for that?) or `:vars` command was called

File: jupyter-lib/api/src/main/kotlin/org/jetbrains/kotlinx/jupyter/api/VariableState.kt (hunk 3 above)
Comment (nikolay-egorov) [reply jupyter-lib/api/src/main/kotlin/org/jetbrains/kotlinx/jupyter/api/VariableState.kt:44]: Indeed. didn't know about it

File: jupyter-lib/api/src/main/kotlin/org/jetbrains/kotlinx/jupyter/api/VariableState.kt (hunk 8 above)
Comment (nikolay-egorov) [reply jupyter-lib/api/src/main/kotlin/org/jetbrains/kotlinx/jupyter/api/VariableState.kt:33]: For sure)

**Comment by nikolay-egorov** [comment]
I guess we can leave it as it is for now and just send empty strings in meta data. Might reuse serialization flag we already have

File: jupyter-lib/api/src/main/kotlin/org/jetbrains/kotlinx/jupyter/api/VariableState.kt (hunk 5 above)
Comment (nikolay-egorov) [reply jupyter-lib/api/src/main/kotlin/org/jetbrains/kotlinx/jupyter/api/VariableState.kt:29]: handleIfRecursiveStructure ))

**Pull request merged**